### ISSUE TITLE
About autograd

### ISSUE THREAD

**Issue by zorrocai** [jxgu1016/MNIST_center_loss_pytorch#7]
I noticed that you have written center loss with backward designed by yourself. What would happen if  define the forward function only,and use autograd in pytorch？I wonder whether there lies any differences between them.

**Comment by jxgu1016** [comment]
1. Actually, the backward function is not designed by me but by the author of **center loss** according to the equation 4  in the paper.
2. The difference may be negligible, I think.

**Comment by zorrocai** [comment]
Thanks for reply, so quickly~.
Actually, I mean that you write the backward in the **class Centerloss**.
And I do some change on the code,without the self-defined backward function :
class CenterLoss(nn.Module):
    def __init__(self, num_classes, feat_dim ):
        super(CenterLoss, self).__init__()
        self.centers = nn.Parameter(torch.randn(num_classes, feat_dim))
        #self.centerlossfunc = CenterlossFunc.apply
        self.feat_dim = feat_dim

    def forward(self, label, feat):
        batch_size = feat.size(0)
        feat = feat.view(batch_size, -1)
        # To check the dim of centers and features
        if feat.size(1) != self.feat_dim:
            raise ValueError("Center's dim: {0} should be equal to input feature's dim: {1}".format(self.feat_dim,feat.size(1)))
        
        centers_batch = self.centers.index_select(0, label.long())
        return (feat - centers_batch).pow(2).sum() / 2.0

Guess what happened? 
the loss is NAN, It seems that the derivative function given by the author  is  important somehow.

Training... Epoch = 1
tensor(779.4907, device='cuda:0')
tensor(24004.1328, device='cuda:0')
tensor(1.00000e+05 *
       6.8511, device='cuda:0')
tensor(2.3724e+07, device='cuda:0')
tensor(6.8685e+08, device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')
tensor(nan., device='cuda:0')

**Comment by jxgu1016** [comment]
It's weird and I believe it should work.
You may check my old version of [center loss](https://github.com/jxgu1016/MNIST_center_loss_pytorch/blob/dbeea5380de8a3c6b1b3b3f2c411b980e143dd87/CenterLoss.py).

**Comment by zorrocai** [comment]
Well, It seems that in your
https://github.com/jxgu1016/MNIST_center_loss_pytorch/blob/dbeea5380de8a3c6b1b3b3f2c411b980e143dd87/CenterLoss.py
instead of using <a href="https://www.codecogs.com/eqnedit.php?latex=L_{c}&space;=&space;\frac{1}{2}\sum_{i=1}^{m}\left&space;\|&space;x_{i}-C_{y_{i}}&space;\right&space;\|\tfrac{2}{2}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?L_{c}&space;=&space;\frac{1}{2}\sum_{i=1}^{m}\left&space;\|&space;x_{i}-C_{y_{i}}&space;\right&space;\|\tfrac{2}{2}" title="L_{c} = \frac{1}{2}\sum_{i=1}^{m}\left \| x_{i}-C_{y_{i}} \right \|\tfrac{2}{2}" /></a> from the original paper, you choosed to use <a href="https://www.codecogs.com/eqnedit.php?latex=L_{c}&space;=&space;\frac{1}{2}\sum_{i=1}^{m}\frac{1}{N_{y_{i}}}\left&space;\|&space;x_{i}-C_{y_{i}}&space;\right&space;\|\tfrac{2}{2}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?L_{c}&space;=&space;\frac{1}{2}\sum_{i=1}^{m}\frac{1}{N_{y_{i}}}\left&space;\|&space;x_{i}-C_{y_{i}}&space;\right&space;\|\tfrac{2}{2}" title="L_{c} = \frac{1}{2}\sum_{i=1}^{m}\frac{1}{N_{y_{i}}}\left \| x_{i}-C_{y_{i}} \right \|\tfrac{2}{2}" /></a>.
What a brilliant change it is!
